protocol/l1-upgrades.md

```diff
@@ -0,0 +1,305 @@
+
+/// @notice forwards calldata to the specified addresses.
+///   Reverts unless called by owner.
+function forward(address[], bytes[]) external;
+```
+
+This architecture makes it possible to manage upgrade authorization for the whole superchain through
+a single storage value in a single contract, rather than having a separate contract per OP Chain.
+
+It also avoids the need for an `sload` to ready the `proxyType` mapping for each contract.
+
+## Identifying OP Chains by the SystemConfig
+
+Rather than tracking the contracts within the `ProxyAdmin`, we will use the `SystemConfig` as the
+unique identifier for each OPChain, and get the contract addresses from it. The custom logic
+required for the two non-standard proxy types will be handled in the OPCM contract, and does not
+require tracking the proxy type, as it is known already for each contract, ie. the only non-standard
+Proxies are the `L1StandardBridge` (`L1ChugsplashProxy`) and `L1CrossDomainMessenger`
+(`ResolvedDelegateProxy`).
+
+## The upgrade process
+
+As part of the release process, the associated implementation contracts and a new OPCM, will be deployed.
+
+The Upgrade Controller MUST be a Safe in order to atomically:
+
+1. transfer ownership of the `SuperchainProxyAdmin` to the new OPCM.
+2. call `OPCM.upgrade()` which will return ownership to the Safe immediately upon completing the
+   upgrade.
+
+Thus the high level logic for upgrading a contract should be roughly as follows:
+
+```js
+SuperchainProxyAdmin admin;
+ISystemConfigs[] public systemConfigs;
+
+Addresses immutable implementations;
+
+constructor(Addresses _implementations, ISystemConfigs[] _systemConfigs) {
+  implementations = _implementations;
+  for(uint i=0; i < _systemConfigs.length; i++){
+    systemConfigs.push(_systemConfigs[i])
+  }
+}
+
+function upgrade() public {
+  for(uint i=0; i< systemConfigs.length; i++) {
+    // Read the `Addresses` struct from each `SystemConfig` in the `systemConfigs` mapping.
+    // For each entry in the `Addresses` struct:
+      // 1. Call the `SuperchainProxyAdmin` to update the implementation to the `Populator` contract and call `Populator.upgrade()`.
+      // 2. Update the implementation to the final implementation contract.
+  }
+  // Return ownership of the `SuperchainProxyAdmin` to the Upgrade Controller (`msg.sender`)
+      //    `SuperchainProxyAdmin`. The unique upgrade call encoding required for the non-standard proxies
+      //    should be handled here.
+}
+```
+
+**Note:**A new getter should be added to the `SystemConfig` to perform this in a single call.
+
+## L1ChugsplashProxy Considerations
+
+The L1 Chugsplash proxy is much more expensive to upgrade, as it is done by calling
+[`setCode()`](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/src/legacy/L1ChugSplashProxy.sol#L92)
+which creates a new implementation contract. In this sense the L1StandardBridge is not truly MCP-L1
+ready.
```

Comment:
What the ProxyAdmin currently does for chugsplash proxies is call `setStorage` on the proxy to directly overwrite the implementation address

Reply:
Ahhh, good catch! 

So then what we could do for the `L1StandardBridge` is have the make that storage write at the end `upgrade()`/`initialize()`. 

And thinking about it now, perhaps we should do that for all of the contracts so that: 
1. they all have a consistent upgrade flow
2. we have two fewer CALLs per contract being upgraded
wdyt?

Reply:
I removed this section as no longer relevant.